### ISSUE TITLE
mod-wsgi is not working with python3.7

### ISSUE THREAD

**Issue by abhiabhi94** [GrahamDumpleton/mod_wsgi#505]
I am trying to deploy a django project but I get a `wsgi` error from the server. These files might help in giving more details regarding my setup and file structure.  

**Error log**:
```log
[Fri Dec 27 08:44:02.403611 2019] [wsgi:error] [pid 28373] [remote 192.168.1.104:52718] mod_wsgi (pid=28373): Target WSGI script '/home/ashish/Dev/we$
[Fri Dec 27 08:44:02.403655 2019] [wsgi:error] [pid 28373] [remote 192.168.1.104:52718] mod_wsgi (pid=28373): Exception occurred processing WSGI scri$
[Fri Dec 27 08:44:02.403804 2019] [wsgi:error] [pid 28373] [remote 192.168.1.104:52718] Traceback (most recent call last):
[Fri Dec 27 08:44:02.403824 2019] [wsgi:error] [pid 28373] [remote 192.168.1.104:52718]   File "/home/ashish/Dev/webDev/blog/blog/wsgi.py", line 12, $
[Fri Dec 27 08:44:02.403828 2019] [wsgi:error] [pid 28373] [remote 192.168.1.104:52718]     from django.core.wsgi import get_wsgi_application
[Fri Dec 27 08:44:02.403841 2019] [wsgi:error] [pid 28373] [remote 192.168.1.104:52718] ModuleNotFoundError: No module named 'django'

```
**apache2.conf**
```conf
        WSGIScriptAlias / /home/ashish/Dev/webDev/blog/blog/wsgi.py
        WSGIDaemonProcess django_process python-path=/home/ashish/Dev/webDev/blog python-home=/home/ashish/.virtualenvs/djangoblog
        WSGIProcessGroup django_process

        <Directory /home/ashish/Dev/webDev/blog/assets>
                Require all granted
        </Directory>

        <Directory /home/ashish/Dev/webDev/blog/media>
                Require all granted
        </Directory>

        <Directory /home/ashish/Dev/webDev/blog/blog>
                <Files wsgi.py>
                        Require all granted
                </Files>
        </Directory>

```
I have made the `virtual environment` using the `python3 -m venv` command.  
It(`djangoblog`) has `python3.7` installed.  

I have made `www:data` the owner of both `/home/ashish/Dev/webDev/blog` and `/home/ashish/.virtualenvs/djangoblog`, also given them permissions `755`.  

**Other useful information**:  

*OS*: Ubuntu 18.04.3  
*Apache*: 2.4.29  
*mod-wsgi*: 4.5.17  
*Django* : 2.2.5

While going through some of the test scripts as mentioned in the official documentation of `mod-wsgi` regarding python versions https://modwsgi.readthedocs.io/en/develop/user-guides/checking-your-installation.html#python-installation-in-use  

I see this as the output when test it against my `wsgi.py` file.
```python
sys.path = ['/home/ashish/Dev/webDev/blog', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload']
``` 
In the `console`, inside the `virtualenv`
```python
(djangoblog) ashish@ashish-Lenovo-ideapad-330-15IKB:~/Dev/webDev/blog$ python
Python 3.7.3 (default, Apr  3 2019, 19:16:38) 
[GCC 8.0.1 20180414 (experimental) [trunk revision 259383]] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/home/ashish/.virtualenvs/djangoblog/lib/python3.7/site-packages/_pdbpp_path_hack', '/usr/lib/python37.zip', '/usr/lib/python3.7', '/usr/lib/python3.7/lib-dynload', '/home/ashish/.virtualenvs/djangoblog/lib/python3.7/site-packages'
```

Could it be that `mod-wsgi3` uses `python3.6` but `djangoblog` has `python3.7`, be the cause for this error?
Any help in resolving the error would be great.

**Comment by GrahamDumpleton** [comment]
You can't use mod_wsgi compiled for one Python version with a virtual environment for a different version. See:

https://modwsgi.readthedocs.io/en/develop/user-guides/virtual-environments.html#virtual-environment-and-python-version

So you either have to use Python 3.6 for your code, or install mod_wsgi compiled with Python 3.7.

**Comment by abhiabhi94** [comment]
How could I compile mod_wsgi with python 3.7?

**Comment by GrahamDumpleton** [comment]
You have to uninstall your system package for mod_wsgi first. Then the easiest way is to use the `pip install` method to install it into your Python virtual environment, then manually integrate it into the system Apache. For some information, read:

* https://pypi.org/project/mod-wsgi/

**Comment by abhiabhi94** [comment]
Thanks for the update Graham.
According to the docs,
> You can now run the Apache server with mod_wsgi hosting your Django application by running:

```
python manage.py runmodwsgi
```
 Correct me, if I am going wrong here.Earlier, Apache used to run the mod-wsgi daemon which would run the django project.
Now, after installing `mod-wsgi` via `pip,`does it mean that I don't have to make any changes to my current `conf` file for Apache ? How will the Django project run actually?
Sorry, if I might sound a bit confused but I have been stuck with this issue for last 16 hours ;-p

**Comment by GrahamDumpleton** [comment]
If you only care about local development initially, using that is an easy way of doing it. See:

* http://blog.dscpl.com.au/2015/04/using-modwsgi-express-with-django.html
* http://blog.dscpl.com.au/2015/04/integrating-modwsgi-express-as-django.html

**Comment by abhiabhi94** [comment]
No, I do have to take the application to production now.

**Comment by GrahamDumpleton** [comment]
Did you read right down to last section called 'Connecting into Apache installation' in:

* https://pypi.org/project/mod-wsgi/

If you still want to use system Apache, but with mod_wsgi compiled using `pip install`, you need to use method it explains there.

**Comment by abhiabhi94** [comment]
No, I think running the command
```
python manage.py runmodwsgi
```
starts a server that can be used for production, and for serving static files. It is currently not looking  for correct path to the wsgi file, it is looking at `blog/wsgi.py` when it should be looking at `blog/blog/wsgi.py`.  
No issues, I think I should be able to swim through now on my own now. Thanks for the assistance Graham.

**Comment by GrahamDumpleton** [comment]
That is usually because you restructured ``settings.py`` so things were in an additional sub directory, but you didn't fix how `BASE_DIR` was calculated in the now relocated ``settings.py`` files.

**Comment by abhiabhi94** [comment]
I did restructure the `settings.py`.
Currently, I have:
 .
    │   manage.py
    ├───data
    └───blog(django-project)
        ├───settings
        │   │   __init__.py
        │   │   base.py
        │   │   dev.py
        │   │   prod.py
        │   urls.py
        │   wsgi.py
I did actually modify the `BASE_DIR` as 
```
BASE_DIR = os.path.dirname(os.path.dirname(
    os.path.dirname(os.path.abspath(__file__))))
```
so, now `BASE_DIR = /home/ashish/Dev/webDev/blog` which should be okay?

**Comment by GrahamDumpleton** [comment]
It should be looking for `blog/wsgi.py` as when you run `manage.py` the current working directory should be the directory that `manage.py` is in, or conversely, Python would add that directory to `sys.path` if `manage.py` is run from another directory with a path.

What is ``WSGI_APPLICATION`` set to in the settings file? Are you supply the `--working-directory` option to `runmodwsgi`?

Has any code of the application changing the current working directory?

FWIW, you can see how it is worked out  by looking at:

* https://github.com/GrahamDumpleton/mod_wsgi/blob/develop/src/server/management/commands/runmodwsgi.py#L65

**Comment by abhiabhi94** [comment]
```
WSGI_APPLICATION = 'blog.wsgi.application'
```
No i did not use the `--working-directory` option.

**Comment by abhiabhi94** [comment]
Inside the terminal I see
> --working-directory DIRECTORY-PATH
                        Specify the directory which should be used as the
                        current working directory of the WSGI application.
                        This directory will be searched when importing Python
                        modules so long as the WSGI application doesn't
                        subsequently change the current working directory.
When I supply the  path for `wsgi` file explicitly as `/home/ashish/Dev/webDev/blog/blog/wsgi.py`, I 
get an error
```console
[Fri Dec 27 18:18:46.470635 2019] [core:notice] [pid 17482:tid 139688589491072] AH00094: Command line: 'httpd (mod_wsgi-express)   -f /tmp/mod_wsgi-localhost:8000:1000/httpd.conf -D MOD_WSGI_MPM_ENABLE_EVENT_MODULE -D MOD_WSGI_MPM_EXISTS_EVENT_MODULE -D MOD_WSGI_MPM_EXISTS_WORKER_MODULE -D MOD_WSGI_MPM_EXISTS_PREFORK_MODULE -D FOREGROUND'
[Fri Dec 27 18:18:46.470539 2019] [wsgi:alert] [pid 17490:tid 139688589491072] (20)Not a directory: mod_wsgi (pid=17490): Unable to change working directory to '/home/ashish/Dev/webDev/blog/blog/wsgi.py'.
[Fri Dec 27 18:18:46.470651 2019] [wsgi:alert] [pid 17490:tid 139688589491072] mod_wsgi (pid=17490): Failure to configure the daemon process correctly and process left in unspecified state. Restarting daemon process after delay.
```

**Comment by abhiabhi94** [comment]
At this point, I'm thinking whether switching back to Python 3.6 would have probably would have been a better option in the first place. :)

**Comment by abhiabhi94** [comment]
As a matter if I set `DEBUG = TRUE`, the page loads but the static and media files aren't served.
When I set `DEBUG = False`, I get an error in the log:
```
Command line: 'httpd (mod_wsgi-express)   -f /tmp/mod_wsgi-localhost:8000:1000/httpd.conf -D MOD_WSGI_MPM_ENABLE_EVENT_MODULE -D MOD_WSGI_MPM_EXISTS_EVENT_MODULE -D MOD_WSGI_MPM_EXISTS_WORKER_MODULE -D MOD_WSGI_MPM_EXISTS_PREFORK_MODULE -D FOREGROUND

```

**Comment by GrahamDumpleton** [comment]
When I say change the working directory, I am not talking about changing `BASE_DIR`, I am talking about the code executing `os.chdir()` to change the current working directory of the process. This can break things if running `manage.py` from the same directory because in that case `.` is added to the Python module search path so packages can be found, not an absolute path, so changing the current working directory for the process means packages cannot then be found correctly later when Apache is started up by the `runmodwsgi` command.

That said, your debug output above says `sys.path` had:

```
'/home/ashish/Dev/webDev/blog'
```

in it, which means there is no reason `blog/wsgi.py` shouldn't have been able to be imported. But then I don't know where you added code to show that, and it could be unrelated to `runmodwsgi` command.

As the name implies, the `--working-drectory` option needs to be given a directory, not a filename path. Use it as:

```
--working-directory /home/ashish/Dev/webDev/blog
```

This can workaround problems caused by not getting a setting like `BASE_DIR` correct, or inheriting a a really old Django code base where certain settings relied upon are missing. One of the things it does is add the directory to `sys.path`, but then your debug output says it should already be there.

So when you claim `blog/wsgi.py` cannot be found, can you supply the full error messages which you claim indicates that. You haven't so far and I am basing things off you interpreting what error messages you saw correctly.

Finally, that message in the logs is not an error and just information only. Just because how Apache works, it shoves an `error` prefix on anything coming from the `mod_wsgi` side.

**Comment by abhiabhi94** [comment]
After I purged the `libapache2-mod-wsgi-py3` package, I do not get any message in the `error_log` even though I get an error 500 on the browser. The above message that you told as information just gets displayed when the server was restarted for the first time. As I mentioned earlier, the page loads if I set `DEBUG = True`, but static files aren't loaded(some permission issues probably).

**Comment by abhiabhi94** [comment]
Regarding the `blog/wsgi.py` file not found issue, it was just logged once in the error file. I don't completely remember the error message. Now, that I restarted my machine today, it won't obviously be present inside the log.

**Comment by GrahamDumpleton** [comment]
For lack of static files when using `runmodwsgi`, have you run:

```
python manage.py collectstatic
```

to have all static assets copied into the correct place?

What are `STATIC_ROOT` and `STATIC_URL` set to in the Django settings files?

When you run `runmodwsgi`, it generated stuff into a directory in `/tmp`. At the start of the `apachectl` script in that directory, there is a comment giving the full command line used for `mod_wsgi-express`. What does that say?

**Comment by abhiabhi94** [comment]
That explains the reason the static files weren't shown, since I didn't run the `collectstatic` command. I thought it would be a permission issue because the developer console showed error  403.

After I used the `log-to-terminal` flag, `apachectl` at the start of the script has
```
# ['manage.py', 'runmodwsgi']
```

**Comment by abhiabhi94** [comment]
You were right, the web application run as expected after I run the `collectstatic` command.
Quite weirdly though, after I ran the `collectstatic` command, even when I set `DEBUG = False` I can see all the contents of the web application in the browser except the `media` files. Earlier it was showing error 500.

**Comment by abhiabhi94** [comment]
The reason browser showed error 500 but there was no error in the log file, was due to a line in my `settings.py`
```
STATICFILES_STORAGE = 'django.contrib.staticfiles.storage.ManifestStaticFilesStorage'
```
I had used this to force reloading of static resources on change.
I also learnt a lesson to use `logging` inside a `django` project. I wonder whether porting the code from python 3.7 to 3.6 could have save a lot more time for both of us.
Thanks, Graham for taking time out to help me figure out the issue.
